decryptor/postgresql/pg_decryptor.go

```diff
@@ -523,54 +557,66 @@ func (proxy *PgProxy) ProxyDatabaseConnection(errCh chan<- error) {
 		}
 		proxy.clientConnection.SetWriteDeadline(time.Now().Add(network.DefaultNetworkTimeout))
 
-		if !packetHandler.IsDataRow() {
-			if err = packetHandler.sendPacket(); err != nil {
-				logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorNetworkWrite).WithError(err).Errorln("Can't forward packet")
-				errCh <- err
-				return
-			}
-			timer.ObserveDuration()
-			continue
-		}
-
-		logger.Debugln("Matched data row packet")
-		if err = packetHandler.parseColumns(); err != nil {
-			logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCodingPostgresqlCantParseColumnsDescription).WithError(err).Errorln("Can't parse columns in packet")
+		// Massage the packet. This should not normally fail. If it does, the client will not receive the packet.
+		err := proxy.handleDatabasePacket(packetCtx, packetHandler, logger)
+		if err != nil {
 			errCh <- err
 			return
 		}
 
-		if packetHandler.columnCount == 0 {
-			if err = packetHandler.sendPacket(); err != nil {
-				logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorNetworkWrite).WithError(err).Errorln("Can't send packet on column count 0")
-				errCh <- err
-				return
-			}
-			timer.ObserveDuration()
-			continue
-		}
-
-		logger.Debugf("Process columns data")
-		for i := 0; i < packetHandler.columnCount; i++ {
-			column := packetHandler.Columns[i]
-			if column.IsNull() {
-				continue
-			}
-			newData, err := proxy.onColumnDecryption(packetCtx, i, column.GetData())
-			if err != nil {
-				logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorGeneral).WithError(err).Errorln("Error on column data processing")
-				errCh <- err
-				return
-			}
-			column.SetData(newData)
-		}
-		proxy.decryptor.ResetZoneMatch()
-		packetHandler.updateDataFromColumns()
+		// After tha packet has been observed and possibly modified, forward it to the client.
 		if err = packetHandler.sendPacket(); err != nil {
-			logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorNetworkWrite).WithError(err).Errorln("Can't send packet")
+			logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorNetworkWrite).
+				WithError(err).Errorln("Can't send packet")
 			errCh <- err
 			return
 		}
 		timer.ObserveDuration()
 	}
 }
+
+func (proxy *PgProxy) handleDatabasePacket(ctx context.Context, packet *PacketHandler, logger *log.Entry) error {
+	// Let the protocol observer take a look at the packet, keeping note of it.
+	packetType, err := proxy.protocolState.HandleDatabasePacket(packet)
+	if err != nil {
+		return err
+	}
+	switch packetType {
+	case DataPacket:
+		// If that's some sort of a packet with response data inside it, go on...
+
+	default:
+		// Forward all other uninteresting packets to the client without processing.
+		return nil
+	}
+
+	logger.Debugln("Matched data row packet")
+	if err = packet.parseColumns(); err != nil {
```

Comment:
same here with data packet. better to encapsulate to separate function and call it in correct case branch.